app/src/main/java/com/urbandroid/sleep/addon/stats/model/socialjetlag/SocialJetlagStats.java

```diff
@@ -64,17 +75,175 @@ public SocialJetlagStats narrow(Date from, Date to) {
+                nonOverFlowingTimeIntervals.add(splitIntervals[0]);
+                nonOverFlowingTimeIntervals.add(splitIntervals[1]);
+            }else{
+                nonOverFlowingTimeIntervals.add(new Interval(intervalStart, intervalEnd));
+            }
+        }
+
+        return nonOverFlowingTimeIntervals;
+    }
+
+    /**
+     * (1) Store in values awakeTimes as a 1440 (24 hrs x 60 min) sized bitset
+     *     such that each bitset value is non-empty if in "awake-state"
+     *     0 if in "sleep-state"
+     * (2) Key is the date in String format yyyy-MM-dd
+     * (3) Used for easy cross-day comparison of intersecting time intervals
+     * @param awakeTimes List of intervals that indicate user "awake" state
+     * @return sleepStateByDateMap
+     */
+    public TreeMap<String, BitSet> createSleepStateByDateMap(List<Interval> awakeTimes){
+        TreeMap<String, BitSet> sleepStateByDateMap = new TreeMap<>();
+        for(int j = 0; j < awakeTimes.size(); ++j){
+            Interval awakeInterval = awakeTimes.get(j);
+
+            BitSet sleepState;
+
+            DateTime awakeIntervalStartTime = awakeInterval.getStart();
+            String awakeIntervalFrom = awakeIntervalStartTime.toString("yyyy-MM-dd");
+
+            if(sleepStateByDateMap.containsKey(awakeIntervalFrom)){
+                sleepState = sleepStateByDateMap.get(awakeIntervalFrom);
+            } else{
+                sleepState = new BitSet(1440); // 24 hrs * 60 minutes
+            }
+
+            int startIdx = awakeIntervalStartTime.getMinuteOfDay();
+            int endIdx = awakeInterval.getEnd().getMinuteOfDay();
+            sleepState.set(startIdx, endIdx);
+
+            sleepStateByDateMap.put(awakeIntervalFrom, sleepState);
+        }
+
+        return sleepStateByDateMap;
+    }
+
+    /**
+     * Calculates SRI score given 2 days
+     * It is the (total number of minutes in which a user is in a consistent sleep state) / 1440 mins
+     * Consistent state: prevDay[i] == nextDay[i]
+     * Inconsistent state: prevDay[i] =/= nextDay[i]
+     * @param prevDay
+     * @param nextDay
+     * @return
+     */
+    public float calculateSRI(BitSet prevDay, BitSet nextDay){
+        BitSet minsInconsistentSleepState = (BitSet) prevDay.clone();
+        minsInconsistentSleepState.andNot(nextDay); // store inconsistent sleep state mins in prevDayTmp
+
+        return (1.0f - (minsInconsistentSleepState.length()/1440.f)); // 1 - totalMinsOfInconsistentSleepState
+    }
+
+    /**
+     * Calculates average SRI across multiple, continuous dates
+     * @param sleepStateByDateMap
+     * @return
+     */
+    public float calculateAverageSRI(TreeMap<String, BitSet> sleepStateByDateMap){
+
+        float cumulativeSRI = 0;
+
+        Set keys = sleepStateByDateMap.keySet();
+
+        Iterator nextDayIt = keys.iterator();
+        // iterate through pairs of days to calculate the SRI
+        if (nextDayIt.hasNext()) {
```

Comment:
use of iterator makes sense for pairs of days, but would a for loop be more readable and reduce the need for typecasting to string?

Reply:
After testing, it seems that this does not handle the case when only one element is provided. Instead, it bypasses the for loop and returns nan

Reply:
i think that should be the expected case, because when only one date is provided, you can't calculate the SRI - you need at least two dates to calculate a value

Reply:
the typecast to the string is my unfortunate design of storing dates "conveniently" -  maybe we should discuss in a later commit to refactor